### ISSUE TITLE
v2.0.0 Roadmap

### ISSUE THREAD

**Issue by caseyjhol** [snapappointments/bootstrap-select#2228]
- [ ] drop support for IE < 11
- [ ] make `'auto'` the default value of `dropdownAlignRight`
- [ ] remove `dropupAuto` and replace with `dropup` where `true`, `false`, and `'auto'` (default) are options
- [ ] remove support for `show-menu-arrow`
- [ ] Replace existing title functionality with placeholder. Allow the title attribute to be dynamically generated (like it is currently) or set to static value
- [ ] change default width setting to `'fit'` instead of 220px if form-control class not set
- [ ] change default `selectedTextFormat` setting to `'count'`
- [ ] remove `data-hidden` support for options (can simply use the native `hidden` instead)
- [ ] replace `mobile` method with a mobile-optimized version of bootstrap-select, to support all of bootstrap-select's functionality
- [ ] rename old `mobile` method to something like `useNative` (open to suggestions)
- [ ] replace `actionsBox` with something more robust (#1299). Possible usage could be `buttons: ['selectAll', 'disableAll']`. Allow way to specify location - top or bottom of menu. I'd also like to add the ability to create custom buttons with custom text/function, but I'm still deciding on whether or not to separate this out into another option, such as `defineButtons`, or have it all under one option.
- [ ] use Bootstrap 4 formatting by default
- [ ] allow `liveSearch` to be set to an integer (to automatically enable it if above x number of options). default to 10
- [ ] improved templating options (allow option customize the template for everything, including the option to use a function)
- [ ] selectpicker should refresh itself on changes instead of needing to be told (#1360)
- [ ] replace `data-style` with something more substantial/less confusing. At the very least a rename to data-class might make more sense.

**Comment by felixvictor** [comment]
Good to see further development.

What about future bootstrap v5 support (ie. jquery removal)? https://github.com/twbs/bootstrap/pull/23586

**Comment by caseyjhol** [comment]
I definitely want to support Bootstrap 5 and offer a version without jQuery as a dependency. I've already begun removing jQuery code and replacing it with native JS where possible. It'll be much easier to support/maintain once IE < 11 support is dropped, since we won't be as limited and will be able to remove most of the polyfills. I'm thinking that might be a more realistic goal for bootstrap-select v3.0.0, though.

**Comment by Amunak** [comment]
Can you please try to make V2.0 easier to (re-)style? For starters, allowing us to edit what is now `elementTemplates` directly through configuration would be great.

I have a fairly specific use case where I not only use a tick icon - I have an "icon" for both "selected" and "empty" state, and they differ from multiselect:

![firefox_2019-05-07_21-27-41](https://user-images.githubusercontent.com/781546/57327328-fbd48300-710e-11e9-86ce-6d2cc1ddde42.png)

to regular select:

![firefox_2019-05-07_21-27-27](https://user-images.githubusercontent.com/781546/57327338-0262fa80-710f-11e9-8ab1-68222df3896b.png)

Basically I wanted to make it look like there are checkboxes and radio buttons inside. Currently this is only possible by *patching the source*, which is less then ideal.

**Comment by Organizer21** [comment]
One of the reasons I still like Select2 is the option to easy format inputs to show selected rows as tags... is that something possible to add to an improved version of bootstrap select?

**Comment by 7iomka** [comment]
@Amunak - please, please tell me how you implemented what you showed in the screenshots?
I now have almost such a task

**Comment by Amunak** [comment]
@7iomka sorry, "you don't want to know". I use some ugly patches that directly modify Bootstrap-Select's source when it's pulled by `npm install`. They break basically every time BS updates.

Because of the issues styling BS and others (like what I reported in #2272) I'm actually considering just abolishing this plugin altogether, either making my own that actually suits me or trying to find out something else :/.

**Comment by 7iomka** [comment]
> @7iomka sorry, "you don't want to know". I use some ugly patches that directly modify Bootstrap-Select's source when it's pulled by `npm install`. They break basically every time BS updates.
> 
> Because of the issues styling BS and others (like what I reported in #2272) I'm actually considering just abolishing this plugin altogether, either making my own that actually suits me and doesn't break stuff with every release or trying to find out something else :/.

I'm used another plugin for do that (bootstrap-multiselect)
But after that, I go back justo for experiments and this is result: https://codepen.io/7iomka/pen/xxxZpwN

**Comment by kev22257** [comment]
Would it be possible instead of allowing liveSearch to be an integer to instead (or additionally) be a function?

"allow liveSearch to be set to an integer (to automatically enable it if above x number of options). default to 10"

**Comment by ts1985** [comment]
> I definitely want to support Bootstrap 5 and offer a version without jQuery as a dependency. I've already begun removing jQuery code and replacing it with native JS where possible. It'll be much easier to support/maintain once IE < 11 support is dropped, since we won't be as limited and will be able to remove most of the polyfills. I'm thinking that might be a more realistic goal for bootstrap-select v3.0.0, though.

Is there any timeline for this? (or first a timeline for v2.0.0?)

**Comment by mrlife** [comment]
> > I definitely want to support Bootstrap 5 and offer a version without jQuery as a dependency. I've already begun removing jQuery code and replacing it with native JS where possible. It'll be much easier to support/maintain once IE < 11 support is dropped, since we won't be as limited and will be able to remove most of the polyfills. I'm thinking that might be a more realistic goal for bootstrap-select v3.0.0, though.
> 
> Is there any timeline for this? (or first a timeline for v2.0.0?)

@caseyjhol We love this component, because it's the best Bootstrap select component available open source. 🚀

Could you comment a little on the future of this repository/component and whether it is still being maintained? Many thanks!

**Comment by mrlife** [comment]
> > I definitely want to support Bootstrap 5 and offer a version without jQuery as a dependency. I've already begun removing jQuery code and replacing it with native JS where possible. It'll be much easier to support/maintain once IE < 11 support is dropped, since we won't be as limited and will be able to remove most of the polyfills. I'm thinking that might be a more realistic goal for bootstrap-select v3.0.0, though.
> 
> Is there any timeline for this? (or first a timeline for v2.0.0?)

@caseyjhol We love this component, because it's the best open source Bootstrap select component available. 🚀

Could you comment a little on the future of this repository/component and whether it is still being maintained? Many thanks!

Here is the Bootstrap release schedule, with Bootstrap 5 scheduled for December 2020. https://github.com/twbs/release#upcoming-release-schedule

**Comment by caseyjhol** [comment]
I've been absolutely swamped with work as of late, but I'm hoping to get back into continuing maintenance of bootstrap-select very soon. My priorities are:

1. Finalize #899. Or, at the very least get it to the point where that feature is in "beta" and previous functionality is not broken.
2. Add Bootstrap 5 support.
3. Begin immediate work on dropping jQuery requirement.

**Comment by teropes** [comment]
Feature idea: option to set prefix string or "label" for the select element. It would allow getting rid of external labels that describe what the selection is about. It would be used on both in case of non selected and also when there are values selected.

For example instead of having:
Status: [ None selected ]
Status: [ Queued, Running ]
Status: [ 3 of 6 selected ]

I would like to have string prefix in the input itself:
[ Status: none selected ]
[ Status: Queued, Running ]
[ Status: 3 of 6 selected ]

My value for the option in this case would be "Status: ".

Reasoning for me is that when having inline filter options for some list of data, it takes extra space. In case of typical Search input text box, I might not want to use separate label like Search: [                       ], instead I use the input placeholder option and have just [ Search ] so it would be similar UX in these select boxes.

**Comment by caseyjhol** [comment]
With v1.14.0-beta released, I think I might start work on supporting Bootstrap 5 prior to dropping jQuery as that will most likely require more of an overhaul. While not ideal, you could still run jQuery along with Bootstrap 5 in the meantime.

**Comment by zorn-v** [comment]
> While not ideal, you could still run jQuery along with Bootstrap 5 in the meantime.

No, in bootstrap 5 `data` attributes was changed with prefix `bs-`, eg `data-bs-toggle="dropdown"`

**Comment by samihoda** [comment]
Any news? I was really hoping for BS5 support.

**Comment by bjsintay** [comment]
We still use jQuery and BS5. Would be happy with an interim solution that required both.

**Comment by caseyjhol** [comment]
Bootstrap 5 support has been added in [v1.14.0-beta2](https://github.com/snapappointments/bootstrap-select/releases/tag/v1.14.0-beta2)! jQuery is still required.

**Comment by zorn-v** [comment]
Is help needed to drop jquery ?

**Comment by JshGrn** [comment]
Any updates on the drop of jQuery?

**Comment by NicolasCARPi** [comment]
@caseyjhol what do you think about releasing a beta3 version? Will also be the opportunity to test the new release process!

**Comment by caseyjhol** [comment]
@NicolasCARPi I'm still working on resolving #2659. It seems there are still some issues with that functionality. Hoping to get it sorted out soon, though. I'm definitely excited to test out the new release process as well.

**Comment by jayden2** [comment]
Any chance to support bootstrap 5's `dropdown-menu-dark` class?
https://getbootstrap.com/docs/5.1/components/dropdowns/#dark-dropdowns

**Comment by mrlife** [comment]
@NicolasCARPi @caseyjhol Any news you can share on 1.14 or 2.0 release?

**Comment by AgentSmith0** [comment]
Any updates to this project?

**Comment by NicolasCARPi** [comment]
Hello everyone.

Unfortunately, it seems developments on this project have come to a halt. I tried to impulse a new dynamic but without success to be honest.

bootstrap-select accumulated a lot of technical debt and grew its number of features, which makes it something hard to maintain due to the large single file codebase. I would not consider it a dead project, but it is not active either (it is a superposition of both states). If you browse the code you'll find a lot of things that can easily be removed now that browser support is here. But this is not the path chosen by the maintainer who wishes to keep backward compatibility. The project is also missing a regular release process.

At this point I would recommend considering other options or following the [Three F's of Open Source](https://boyter.org/posts/the-three-f-s-of-open-source/).

I have found these projects:

* https://react-select.com/home
* https://github.com/downshift-js/downshift#readme

As for me and my project, I started working on a library doing what I need so I can replace bootstrap-select when it's done.

**Comment by carstenfuchs** [comment]
This is very very unfortunate. In addition to the above, I've found these alternatives:

  - https://select2.org/
  - https://choices-js.github.io/Choices/
  - https://tom-select.js.org/

These appear to be modern and well maintained – however, bootstrap-select is still the one that works best for me.

(For example, most multi-select implementations remove the selected items from the dropdown list when they put them in the pills list. This is not viable for my use case where I must keep selected items also in the dropdown list with a checkmark next to them. bootstrap-select is afaiks the only library that does that. This in turn is what keeps me stuck with Bootstrap 3 … :disappointed: )

Personally, I would very much appreciate if @caseyjhol reconsidered and kept bootstrap-select active and up to date. I Iack the JavaScript skills to come up with the required changes myself and so can only, but would be willing to, provide indirect means of support.

**Comment by Amunak** [comment]
Thank you @NicolasCARPi for being upfront and honest. Things happen it's not easy to keep "hobby" projects running for years, so it's good to know when people should perhaps look elsewhere.

**Comment by ghost** [comment]
Hello everyone!

I would like to contribute to this open project. 

*But from what I noticed some people are leaving here for another open project.*

**Comment by atanbal** [comment]
hye

**Comment by AgentSmith0** [comment]
If anyone is interested in an actively maintained alternative to bootstrap-select, they can have a look at this: https://github.com/orchidjs/tom-select

**Comment by ericplan** [comment]
Bootstrap-select 2.0.0-beta1 is available via CDN but without documentation.
https://cdnjs.com/libraries/bootstrap-select/2.0.0-beta1
I tried this version, but apart from showing the button **with icon** it did nothing, no error, no dropdown.
Someone got tips to try to get it to work with Bootstrap 5.3.0? As said, no errors, Javascript console is empty.

Same configuration works ok with version 1.13.18, but **without the icon on the button**.

**Comment by extraric** [comment]
> Bootstrap-select 2.0.0-beta1 is available via CDN but without documentation. https://cdnjs.com/libraries/bootstrap-select/2.0.0-beta1

As said before, the CDN 2.0.0 links are actually v1.6.3 (2015), just look at the file source..

**Comment by Bruskyer** [comment]
In the previous version, there may be problems when doing live-search in different languages. (iİçÇşŞüÜğĞöÖ) We have to change the transformation of these letters with code during the search.

**Comment by bairog** [comment]
@NicolasCARPi Any news about your library that aimed to replace bootstrap-select? Is it ready for testing? Thank you.

**Comment by NicolasCARPi** [comment]
@bairog We're now using [TomSelect](https://tom-select.js.org/), it's pretty good and fit our needs to replace bootstrap-select.

**Comment by smoore-srcity** [comment]
Another option for those looking for an alternative to bootstrap-select is [SlimSelect](https://slimselectjs.com)   I've looked at all the other ones mentioned above, including TomSelect.   SlimSelect seems the most similar to what we do with bootstrap-select.   (especially the built in support for some data attributes at the option level)
